app/services/permissions/order.rb

```diff
@@ -28,27 +36,35 @@ def visible_line_items
 
     # Any line items that I can edit
     def editable_line_items
-      Spree::LineItem.where(order_id: editable_orders)
+      Spree::LineItem.where(order_id: editable_orders.select(:id))
     end
 
     private
 
     # Any orders placed through any hub that I manage
-    def managed_orders
-      Spree::Order.where(distributor_id: @permissions.managed_enterprises.select("enterprises.id"))
+    def managed_orders_where_values
+      Spree::Order.
+        where(distributor_id: @permissions.managed_enterprises.select("enterprises.id")).
```

Comment:
why don't we `inner join` orders with `@permissions.managed_enterprises`? By checking the scope as is defined in `Enterprise` I don't see why we need the DB to go fetch the ids, even if within the same query. I'd check if doing so brings any improvement.

Reply:
ah, this is a really good clue!!! but this condition is placed within a series of clauses with ANDs and ORs...
So, instead of
```sql
OR spree_orders.distributor_id IN
           (SELECT enterprises.id
            FROM enterprises
            INNER JOIN enterprise_roles ON enterprise_roles.enterprise_id = enterprises.id
            WHERE (enterprise_roles.user_id = 204238) ))

```
we could do
```sql
INNER JOIN enterprises ON enterprises.id = spree_orders.distributor_id
INNER JOIN enterprise_roles ON enterprise_roles.enterprise_id = enterprises.id
  WHERE (enterprise_roles.user_id = 204238)
```
but the other conditions would not work because this is just one of the conditions, the user does not have to be the manager of the enterprise in all cases...

we could do just for enterprises because the distributor is always there:
```sql
INNER JOIN enterprises ON enterprises.id = spree_orders.distributor_id
WHERE (enterprises.is_primary_producer = 't'
     AND enterprises.id IN ( SELECT enterprise_id FROM enterprise_roles
                                           WHERE (enterprise_roles.user_id = 204238) ))
```

but on top of this there's the super admin condition...

Reply:
maybe you have a clear idea about how the activerecord code should look like?